File: app/services/day_close.py
from collections import Counter
from datetime import date, datetime

from sqlmodel import Session, select

from app.models import Booking, BookingStatus, DailySummary, Patient
from app.time_utils import clinic_today, now_utc, to_clinic_tz


class DayCloseError(ValueError):
    """Tried to close a day that's already closed, or another invariant violation."""


def get_summary(db: Session, clinic_id: int, on: date) -> DailySummary | None:
    return db.exec(
        select(DailySummary).where(
            DailySummary.clinic_id == clinic_id, DailySummary.date == on
        )
    ).first()


def compute_summary(db: Session, clinic_id: int, on: date) -> DailySummary:
    bookings = db.exec(
        select(Booking).where(Booking.clinic_id == clinic_id, Booking.date == on)
    ).all()

    completed = [b for b in bookings if b.status == BookingStatus.done]
    no_shows = sum(1 for b in bookings if b.status == BookingStatus.no_show)
    cancellations = sum(1 for b in bookings if b.status == BookingStatus.cancelled)

    waits: list[int] = []
    consults: list[int] = []
    started_times: list[datetime] = []
    completed_times: list[datetime] = []

    for b in completed:
        if b.checked_in_at and b.started_at:
            waits.append(int((b.started_at - b.checked_in_at).total_seconds()))
        if b.started_at and b.completed_at:
            consults.append(int((b.completed_at - b.started_at).total_seconds()))
            started_times.append(b.started_at)
            completed_times.append(b.completed_at)

    avg_wait = int(sum(waits) / len(waits)) if waits else None
    avg_consult = int(sum(consults) / len(consults)) if consults else None

    if started_times:
        # Peak hour in clinic local time
        hours = Counter(to_clinic_tz(t).hour for t in started_times)
        peak_hour = hours.most_common(1)[0][0]
    else:
        peak_hour = None

    summary = get_summary(db, clinic_id, on)
    if summary is None:
        summary = DailySummary(clinic_id=clinic_id, date=on)
    summary.total_bookings = len(bookings)
    summary.completed = len(completed)
    summary.no_shows = no_shows
    summary.cancellations = cancellations
    summary.avg_wait_seconds = avg_wait
    summary.avg_consult_seconds = avg_consult
    summary.peak_hour = peak_hour
    summary.first_consult_at = min(started_times) if started_times else None
    summary.last_consult_at = max(completed_times) if completed_times else None
    summary.generated_at = now_utc()
    return summary


def close_day(db: Session, clinic_id: int, on: date | None = None) -> DailySummary:
    """Idempotent: re-running on a closed day refreshes the summary stats but
    won't sweep additional patients (none are eligible by then)."""
    on = on or clinic_today()
    cutoff = now_utc()

    # Sweep any still-active bookings to no_show
    candidates = db.exec(
        select(Booking).where(
            Booking.clinic_id == clinic_id,
            Booking.date == on,
            Booking.status.in_([BookingStatus.booked, BookingStatus.checked_in, BookingStatus.in_consult]),
        )
    ).all()
    for b in candidates:
        if b.status == BookingStatus.in_consult:
            # Treat an unfinished in-consult booking as completed for the books
            b.status = BookingStatus.done
            b.completed_at = cutoff
        else:
            b.status = BookingStatus.no_show
            b.no_show_at = cutoff
            patient = db.get(Patient, b.patient_id)
            if patient:
                patient.no_show_count = (patient.no_show_count or 0) + 1
                db.add(patient)
        b.updated_at = cutoff
        db.add(b)

    summary = compute_summary(db, clinic_id, on)
    summary.closed_at = summary.closed_at or cutoff
    db.add(summary)
    db.flush()
    return summary

File: frontend/lib/services/day-close.ts
// Day-close — port of app/services/day_close.py.
//
// Sweeps any still-active bookings to no_show (in_consult is treated as
// `done` for the books — the consult was in progress when the day closed),
// recomputes a daily_summaries row, and stamps `closed_at`. Idempotent —
// running again on a closed day refreshes the stats but doesn't double-sweep.
import { and, eq, inArray, sql } from "drizzle-orm";
import { db, schema } from "@/lib/db/client";
import { clinicToday, nowUtc } from "@/lib/time";
import type { DailySummary, Booking } from "@/lib/db/schema";

const CLINIC_TZ = process.env.CLINIC_TZ ?? "Asia/Kolkata";

export class DayCloseError extends Error {}

export async function getSummary(
  clinicId: number,
  on: string,
): Promise<DailySummary | undefined> {
  const [row] = await db
    .select()
    .from(schema.dailySummaries)
    .where(
      and(
        eq(schema.dailySummaries.clinicId, clinicId),
        eq(schema.dailySummaries.date, on),
      ),
    )
    .limit(1);
  return row;
}

function localHour(d: Date): number {
  return Number(
    new Intl.DateTimeFormat("en-GB", {
      timeZone: CLINIC_TZ,
      hour: "2-digit",
      hour12: false,
    }).format(d),
  );
}

export async function closeDay(
  clinicId: number,
  on?: string,
): Promise<DailySummary> {
  const date = on ?? clinicToday();
  const now = nowUtc();

  // 1. Sweep still-active bookings.
  const candidates = await db
    .select()
    .from(schema.bookings)
    .where(
      and(
        eq(schema.bookings.clinicId, clinicId),
        eq(schema.bookings.date, date),
        inArray(schema.bookings.status, ["booked", "checked_in", "in_consult"]),
      ),
    );

  for (const b of candidates) {
    if (b.status === "in_consult") {
      await db
        .update(schema.bookings)
        .set({ status: "done", completedAt: now, updatedAt: now })
        .where(eq(schema.bookings.id, b.id));
    } else {
      await db
        .update(schema.bookings)
        .set({ status: "no_show", noShowAt: now, updatedAt: now })
        .where(eq(schema.bookings.id, b.id));
      await db
        .update(schema.patients)
        .set({ noShowCount: sql`${schema.patients.noShowCount} + 1` })
        .where(eq(schema.patients.id, b.patientId));
    }
  }

  // 2. Recompute summary from final state.
  const bookings = await db
    .select()
    .from(schema.bookings)
    .where(and(eq(schema.bookings.clinicId, clinicId), eq(schema.bookings.date, date)));

  const completed: Booking[] = [];
  let noShows = 0;
  let cancellations = 0;
  for (const b of bookings) {
    if (b.status === "done") completed.push(b);
    else if (b.status === "no_show") noShows += 1;
    else if (b.status === "cancelled") cancellations += 1;
  }

  const waits: number[] = [];
  const consults: number[] = [];
  const startedTimes: Date[] = [];
  const completedTimes: Date[] = [];
  for (const b of completed) {
    if (b.checkedInAt && b.startedAt) {
      waits.push((b.startedAt.getTime() - b.checkedInAt.getTime()) / 1000);
    }
    if (b.startedAt && b.completedAt) {
      consults.push((b.completedAt.getTime() - b.startedAt.getTime()) / 1000);
      startedTimes.push(b.startedAt);
      completedTimes.push(b.completedAt);
    }
  }

  const avgWait = waits.length
    ? Math.round(waits.reduce((a, b) => a + b, 0) / waits.length)
    : null;
  const avgConsult = consults.length
    ? Math.round(consults.reduce((a, b) => a + b, 0) / consults.length)
    : null;

  let peakHour: number | null = null;
  if (startedTimes.length) {
    const buckets = new Map<number, number>();
    for (const t of startedTimes) {
      const h = localHour(t);
      buckets.set(h, (buckets.get(h) ?? 0) + 1);
    }
    peakHour = [...buckets.entries()].sort((a, b) => b[1] - a[1])[0][0];
  }

  const firstConsultAt = startedTimes.length
    ? new Date(Math.min(...startedTimes.map((t) => t.getTime())))
    : null;
  const lastConsultAt = completedTimes.length
    ? new Date(Math.max(...completedTimes.map((t) => t.getTime())))
    : null;

  const existing = await getSummary(clinicId, date);
  const values = {
    clinicId,
    date,
    totalBookings: bookings.length,
    completed: completed.length,
    noShows,
    cancellations,
    avgWaitSeconds: avgWait,
    avgConsultSeconds: avgConsult,
    peakHour,
    firstConsultAt,
    lastConsultAt,
    closedAt: existing?.closedAt ?? now,
    generatedAt: now,
  };

  if (existing) {
    const [updated] = await db
      .update(schema.dailySummaries)
      .set(values)
      .where(eq(schema.dailySummaries.id, existing.id))
      .returning();
    return updated;
  }
  const [created] = await db
    .insert(schema.dailySummaries)
    .values(values)
    .returning();
  return created;
}
